File: backend/models/special_day.py
import uuid
from datetime import date
from sqlalchemy import String, Text, Date, ForeignKey, func, DateTime, Integer, Boolean
from sqlalchemy.orm import Mapped, mapped_column, relationship
from database.connection import Base
from datetime import datetime


class SpecialDay(Base):
    __tablename__ = "special_days"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    date: Mapped[date] = mapped_column(Date, nullable=False)
    type: Mapped[str] = mapped_column(String(50), nullable=False, default="Birthday")  # Birthday, Wedding, Anniversary, Engagement, Festival, Event, Memory, Custom
    is_recurring: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    custom_type: Mapped[str | None] = mapped_column(String(100), nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    gift_ideas: Mapped[str | None] = mapped_column(Text, nullable=True)
    celebration_plans: Mapped[str | None] = mapped_column(Text, nullable=True)
    reminder_notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    message_draft: Mapped[str | None] = mapped_column(Text, nullable=True)
    user_id: Mapped[int] = mapped_column(Integer, ForeignKey("users.id"), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    user = relationship("User", back_populates="special_days")

File: web/src/types/SpecialDay.ts
export type SpecialDayType =
  | "Birthday"
  | "Wedding Anniversary"
  | "Engagement"
  | "Festival"
  | "Meeting"
  | "Achievement"
  | "Personal Memory"
  | "Custom Event"
  | string;

export interface SpecialDay {
  id: number;
  title: string;
  date: string; // YYYY-MM-DD
  type: SpecialDayType;
  is_recurring: boolean;
  custom_type?: string | null;
  notes?: string | null;
  gift_ideas?: string | null;
  celebration_plans?: string | null;
  reminder_notes?: string | null;
  message_draft?: string | null;

  // Contact / Person fields
  recipient_email?: string | null;
  phone_number?: string | null;
  relationship?: string | null;

  // Smart Reminder System fields
  reminder_enabled?: boolean;
  reminder_type?: string | null; // 'on_day' | '1_day' | '3_days' | '1_week' | '2_weeks' | '1_month' | 'custom'
  reminder_value?: number | null;
  reminder_unit?: string | null; // 'days' | 'weeks' | 'months'
  reminder_time?: string | null;

  // Auto Email Wishes fields
  auto_send_email: boolean;
  email_subject?: string | null;
  email_message?: string | null;
  email_send_time?: string | null;
  last_email_sent_at?: string | null;
  last_sent_year?: number | null;
  timezone?: string | null;
  email_status?: string | null;
  email_retry_count?: number;

  // Extended planning fields
  location?: string | null;
  emoji?: string | null;
  event_image?: string | null;
  favorite_color?: string | null;
  checklist?: string | null; // Serialized JSON or newline text
  budget?: string | null;
  links?: string | null;
  attachments?: string | null;

  user_id?: number;
  created_at?: string;
  updated_at?: string;
}

export interface ScheduledEmail {
  id: number;
  recipient_email: string;
  subject: string;
  body: string;
  send_datetime: string;
  timezone: string;
  status: "scheduled" | "sending" | "sent" | "failed" | "cancelled" | string;
  error_message?: string | null;
  retry_count: number;
  user_id: number;
  important_day_id?: number | null;
  created_at: string;
}
